esm/utils/domains.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * @typedef {import('./diag-domain-config.types.js').DomainInfo} DomainInfo
+ */
+
+/**
+ * Extracts `hostname`, `pathname`, `isWildcard` from a given `domain`
+ *
+ * @param {string} domain
+ * @return {{ hostname: string, pathname: string, isWildcard: boolean }}
+ */
+export function parseDomain (domain) {
+  const domainWithHttp = domain.match(/^https?:\/\//) != null ? domain : 'https://' + domain;
+  // With firefox, 'https://*.toto.com' is considered invalid so we strip it off for the test
+  // because we know this part is valid and we want the rest to be sanitized by the URL parser
+  const isWildcard = domain.startsWith('*.');
+
+  if (!isWildcard && domain.includes('*')) {
+    throw new DomainParseError('invalid-wildcard', 'Invalid wildcard format. "*" may only be used as a subdomain');
+  }
+
+  try {
+    const { hostname, pathname } = new URL(domainWithHttp.replace('*.', ''));
+
+    return {
+      hostname,
+      pathname,
+      isWildcard,
+    };
+  }
+  catch {
+    if (domain.length === 0) {
+      throw new DomainParseError('empty', 'Empty domain value');
+    }
+
+    throw new DomainParseError('invalid-format', 'Invalid domain format');
+  }
+}
+
+export class DomainParseError extends Error {
+  /**
+   * @param {'empty'|'invalid-wildcard'|'invalid-format'} code
+   * @param {string} message
+   */
+  constructor (code, message) {
+    super(message);
+
+    this.code = code;
+  }
```

Comment:
Add an optional `cause` parameter:

```suggestion
  constructor (code, message, cause) {
    super(message);

    this.code = code;
    this.cause = cause;
  }
```